### PR TITLE
ProxyAsync returns ProxyError

### PR DESCRIPTION
**Pull request by Kahbazi**
Fixes https://github.com/microsoft/reverse-proxy/issues/962

### CODE REVIEW

File: src/ReverseProxy/Service/Proxy/IHttpProxy.cs [hunk 1]
```diff
@@ -21,7 +21,7 @@ public interface IHttpProxy
         /// <param name="requestOptions">Options for the outgoing request.</param>
         /// <param name="transformer">Request and response transforms. Use <see cref="HttpTransformer.Default"/> if
         /// custom transformations are not needed.</param>
```
Comment (alnikola) [inline src/ReverseProxy/Service/Proxy/IHttpProxy.cs:23]: Please, add a comment about the returned value.

File: src/ReverseProxy/Service/Proxy/IHttpProxyExtensions.cs [hunk 2]
```diff
@@ -15,7 +15,7 @@ public static class IHttpProxyExtensions
         /// <param name="context">The HttpContent to proxy from.</param>
         /// <param name="destinationPrefix">The url prefix for where to proxy the request to.</param>
         /// <param name="httpClient">The HTTP client used to send the proxy request.</param>
```
Comment (alnikola) [inline src/ReverseProxy/Service/Proxy/IHttpProxyExtensions.cs:17]: Please, add a comment about the returned value.

File: src/ReverseProxy/Service/Proxy/IHttpProxyExtensions.cs [hunk 3]
```diff
@@ -31,7 +31,7 @@ public static class IHttpProxyExtensions
...
         /// <param name="requestOptions">Options for the outgoing request.</param>
```
Comment (alnikola) [inline src/ReverseProxy/Service/Proxy/IHttpProxyExtensions.cs:33]: Please, add a comment about the returned value.

**Review by alnikola — COMMENTED** [review]
Could you please explain why this change is needed? Does it improve test coverage or something else?

**Comment by MihaZupan** [comment]
> Could you please explain why this change is needed? Does it improve test coverage or something else?

Discussed in #962, it makes it a lot more discoverable for callers

File: src/ReverseProxy/Service/Proxy/HttpProxy.cs [hunk 4]
```diff
@@ -181,14 +180,14 @@ public HttpProxy(ILogger<HttpProxy> logger, IClock clock)
...
+                    return ProxyError.ResponseHeaders;
                 }
 
                 // :: Step 7-A: Check for a 101 upgrade response, this takes care of WebSockets as well as any other upgradeable protocol.
                 if (destinationResponse.StatusCode == HttpStatusCode.SwitchingProtocols)
                 {
-                    await HandleUpgradedResponse(context, destinationResponse, requestAborted);
-                    return;
+                    var error = await HandleUpgradedResponse(context, destinationResponse, requestAborted);
+                    return error;
```
Comment (MihaZupan) [inline src/ReverseProxy/Service/Proxy/HttpProxy.cs:190]:
```suggestion
                    return await HandleUpgradedResponse(context, destinationResponse, requestAborted);
```

File: src/ReverseProxy/Service/Proxy/HttpProxy.cs [hunk 5]
```diff
@@ -206,8 +205,8 @@ public HttpProxy(ILogger<HttpProxy> logger, IClock clock)
 
                 if (responseBodyCopyResult != StreamCopyResult.Success)
                 {
-                    await HandleResponseBodyErrorAsync(context, requestContent, responseBodyCopyResult, responseBodyException);
-                    return;
+                    var error = await HandleResponseBodyErrorAsync(context, requestContent, responseBodyCopyResult, responseBodyException);
+                    return error;
```
Comment (MihaZupan) [inline src/ReverseProxy/Service/Proxy/HttpProxy.cs:209]:
```suggestion
                    return await HandleResponseBodyErrorAsync(context, requestContent, responseBodyCopyResult, responseBodyException);
```

**Review by MihaZupan — APPROVED** [review]

**Review by Tratcher — APPROVED** [review]
+ others comments

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**